### ISSUE TITLE
Slow performance when using MGLMapView in Scroll View on iOS 10 (fine on iOS 11)

### ISSUE THREAD

**Issue by mattijsf** [mapbox/mapbox-gl-native#10327]
**Platform:** iOS
**Mapbox SDK version:** 3.6.4

### Steps to trigger behavior

 1. Create a `UIScrollView` layout and fill the View Controller's content
 2. Add a `MGLMapView` inside the scrollview set it's height to 200
 3. Add a dummy view with a large height (I used 1000) underneath the MapView to make the scrollview scrollable:
![interface-layout](https://user-images.githubusercontent.com/706368/32222449-95989240-be39-11e7-8789-f7d588b9bced.png)
 4. Start profiler Product > Profile > Time Profiler
 5. Choose a iOS 9 or 10 Target (I used iPhone 5s / iOS 10.2)
 6. Press Record > Wait for the app to start.
 7. Scroll down until the MapView is no longer visible.
 8. Drag the up and down fast / repeatedly keeping the mapview out of screen.

### Expected behavior

No degradation in performance when the map view is scrolling offscreen. And no difference in behavior between iOS 10 and iOS 11.

### Actual behavior

Observing a high load on the UI thread on device running iOS 9 / 10. While this is not the case on iOS 11.

**iOS 10 (Bad performance):**

![profiler-ios10](https://user-images.githubusercontent.com/706368/32223027-fc02be32-be3b-11e7-8d66-0af7410405d9.png)

**iOS 11: (Good performance):**
![profiler-ios11](https://user-images.githubusercontent.com/706368/32223263-f9bc2b30-be3c-11e7-94fa-3345879b059d.png)

**Slow call stack:**
```
  27  2639.0  MapboxTestApp (58352) :0
  26  2628.0  Main Thread  0xe236f :0
  25 libdyld.dylib 2628.0  start
  24 MapboxTestApp 2628.0  main /Users/mattijs/Desktop/MapboxTestApp/MapboxTestApp/ViewController.swift:12
  23 UIKit 2628.0  UIApplicationMain
  22 GraphicsServices 2628.0  GSEventRunModal
  21 CoreFoundation 2628.0  CFRunLoopRunSpecific
  20 CoreFoundation 2627.0  __CFRunLoopRun
  19 CoreFoundation 2570.0  __CFRunLoopDoTimers
  18 CoreFoundation 2567.0  __CFRunLoopDoTimer
  17 CoreFoundation 2564.0  __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__
  16 QuartzCore 2564.0  CA::Display::DisplayLink::dispatch_items(unsigned long long, unsigned long long, unsigned long long)
  15 QuartzCore 2564.0  CA::Display::DisplayLinkItem::dispatch(unsigned long long)
  14 Mapbox 2564.0  -[MGLMapView updateFromDisplayLink]
  13 GLKit 2564.0  -[GLKView _display:]
  12 Mapbox 2502.0  -[MGLMapView glkView:drawInRect:]
  11 Mapbox 2502.0  mbgl::Map::Impl::render(mbgl::View&)
  10 Mapbox 2492.0  mbgl::Painter::render(mbgl::style::Style const&, mbgl::FrameData const&, mbgl::View&, mbgl::SpriteAtlas&)
   9 Mapbox 2362.0  void mbgl::Painter::renderPass<std::__1::__wrap_iter<mbgl::RenderItem const*> >(mbgl::PaintParameters&, mbgl::RenderPass, std::__1::__wrap_iter<mbgl::RenderItem const*>, std::__1::__wrap_iter<mbgl::RenderItem const*>, unsigned int, signed char)
   8 Mapbox 1094.0  mbgl::Painter::renderLine(mbgl::PaintParameters&, mbgl::LineBucket&, mbgl::RenderLineLayer const&, mbgl::RenderTile const&)
   7 Mapbox 1024.0  void mbgl::Program<mbgl::shaders::line, mbgl::gl::Triangle, mbgl::LineLayoutAttributes, mbgl::gl::Uniforms<mbgl::uniforms::u_matrix, mbgl::uniforms::u_width, mbgl::uniforms::u_ratio, mbgl::uniforms::u_gl_units_to_pixels>, mbgl::style::LinePaintProperties>::draw<mbgl::gl::Triangles>(mbgl::gl::Context&, mbgl::gl::Triangles, mbgl::gl::DepthMode, mbgl::gl::StencilMode, mbgl::gl::ColorMode, mbgl::IndexedTuple<mbgl::TypeList<mbgl::uniforms::u_matrix, mbgl::uniforms::u_width, mbgl::uniforms::u_ratio, mbgl::uniforms::u_gl_units_to_pixels>, mbgl::TypeList<mbgl::gl::UniformValue<mbgl::uniforms::u_matrix, std::__1::array<double, 16ul> >, mbgl::gl::UniformValue<mbgl::uniforms::u_width, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_ratio, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_gl_units_to_pixels, std::__1::array<float, 2ul> > > > const&, mbgl::gl::VertexBuffer<mbgl::gl::detail::Vertex<mbgl::gl::Attribute<short, 4ul>, mbgl::gl::Attribute<unsigned char, 4ul> >, mbgl::gl::Indexed> const&, mbgl::gl::IndexBuffer<mbgl::gl::Triangles> const&, mbgl::SegmentVector<mbgl::gl::Attributes<mbgl::attributes::a_pos_normal, mbgl::attributes::a_data<unsigned char, 4ul>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_opacity>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_color>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_gapwidth>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_offset<1ul> >, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_blur> > > const&, mbgl::PaintPropertyBinders<mbgl::TypeList<mbgl::style::LineOpacity, mbgl::style::LineColor, mbgl::style::LineGapWidth, mbgl::style::LineOffset, mbgl::style::LineBlur> > const&, mbgl::style::PaintProperties<mbgl::style::LineOpacity, mbgl::style::LineColor, mbgl::style::LineTranslate, mbgl::style::LineTranslateAnchor, mbgl::style::LineWidth, mbgl::style::LineGapWidth, mbgl::style::LineOffset, mbgl::style::LineBlur, mbgl::style::LineDasharray, mbgl::style::LinePattern>::Evaluated const&, float, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
   6 Mapbox 1023.0  void mbgl::gl::Program<mbgl::gl::Triangle, mbgl::gl::Attributes<mbgl::attributes::a_pos_normal, mbgl::attributes::a_data<unsigned char, 4ul>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_opacity>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_color>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_gapwidth>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_offset<1ul> >, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_blur> >, mbgl::gl::Uniforms<mbgl::uniforms::u_matrix, mbgl::uniforms::u_width, mbgl::uniforms::u_ratio, mbgl::uniforms::u_gl_units_to_pixels, mbgl::InterpolationUniform<mbgl::attributes::a_opacity>, mbgl::InterpolationUniform<mbgl::attributes::a_color>, mbgl::InterpolationUniform<mbgl::attributes::a_gapwidth>, mbgl::InterpolationUniform<mbgl::attributes::a_offset<1ul> >, mbgl::InterpolationUniform<mbgl::attributes::a_blur>, mbgl::uniforms::u_opacity, mbgl::uniforms::u_color, mbgl::uniforms::u_gapwidth, mbgl::uniforms::u_offset, mbgl::uniforms::u_blur> >::draw<mbgl::gl::Triangles>(mbgl::gl::Context&, mbgl::gl::Triangles, mbgl::gl::DepthMode, mbgl::gl::StencilMode, mbgl::gl::ColorMode, mbgl::IndexedTuple<mbgl::TypeList<mbgl::uniforms::u_matrix, mbgl::uniforms::u_width, mbgl::uniforms::u_ratio, mbgl::uniforms::u_gl_units_to_pixels, mbgl::InterpolationUniform<mbgl::attributes::a_opacity>, mbgl::InterpolationUniform<mbgl::attributes::a_color>, mbgl::InterpolationUniform<mbgl::attributes::a_gapwidth>, mbgl::InterpolationUniform<mbgl::attributes::a_offset<1ul> >, mbgl::InterpolationUniform<mbgl::attributes::a_blur>, mbgl::uniforms::u_opacity, mbgl::uniforms::u_color, mbgl::uniforms::u_gapwidth, mbgl::uniforms::u_offset, mbgl::uniforms::u_blur>, mbgl::TypeList<mbgl::gl::UniformValue<mbgl::uniforms::u_matrix, std::__1::array<double, 16ul> >, mbgl::gl::UniformValue<mbgl::uniforms::u_width, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_ratio, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_gl_units_to_pixels, std::__1::array<float, 2ul> >, mbgl::gl::UniformValue<mbgl::InterpolationUniform<mbgl::attributes::a_opacity>, float>, mbgl::gl::UniformValue<mbgl::InterpolationUniform<mbgl::attributes::a_color>, float>, mbgl::gl::UniformValue<mbgl::InterpolationUniform<mbgl::attributes::a_gapwidth>, float>, mbgl::gl::UniformValue<mbgl::InterpolationUniform<mbgl::attributes::a_offset<1ul> >, float>, mbgl::gl::UniformValue<mbgl::InterpolationUniform<mbgl::attributes::a_blur>, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_opacity, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_color, mbgl::Color>, mbgl::gl::UniformValue<mbgl::uniforms::u_gapwidth, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_offset, float>, mbgl::gl::UniformValue<mbgl::uniforms::u_blur, float> > > const&, mbgl::gl::VertexArray&, mbgl::IndexedTuple<mbgl::TypeList<mbgl::attributes::a_pos_normal, mbgl::attributes::a_data<unsigned char, 4ul>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_opacity>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_color>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_gapwidth>, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_offset<1ul> >, mbgl::ZoomInterpolatedAttribute<mbgl::attributes::a_blur> >, mbgl::TypeList<std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding>, std::experimental::fundamentals_v1::optional<mbgl::gl::AttributeBinding> > > const&, mbgl::gl::IndexBuffer<mbgl::gl::Triangles> const&, unsigned long, unsigned long)
   5 GLEngine 1019.0  glDrawElements_IMM_ES2Exec
   4 GLEngine 997.0  gleFlushAtomicFunc
   3  579.0  0x11cb97cf4
   2 GLRendererFloat 577.0  gldRenderFillTriangles
   1  458.0  0x11cb99cd4
   0  28.0  0x11cb9a6b8
```

**Comment by mattijsf** [comment]
For what it's worth, I repeated the test with an instance of Apple's `MKMapView` where the performance is equal on both platforms.

**Comment by mattijsf** [comment]
Does anyone have an idea what is going on and why this is only happening while the map is out of view? Or does someone have a possible workaround for this?

It's a pretty serious performance issue for us so any help is welcome :)

**Comment by iamZoltanVaradi** [comment]
I began facing the same issue after updating to iOS 11. With both v3.5.4 and v3.6.4 the UI is very choppy on iOS 11, but works great on iOS 10. 
For me the Time Profiler says the app spends 99.5% of the CPU time in MGLMapView's UpdateFromDisplayLink function.

In my app, the MGLMapView 's simply sitting in the ViewController's view, not in any scrollviews.

![screen shot 2017-11-09 at 11 12 07](https://user-images.githubusercontent.com/1964469/32600166-022f7b1a-c53f-11e7-8cdf-ea4676d4c62f.png)

**Comment by mattijsf** [comment]
@iamZoltanVaradi Interesting to read. Stack looks very similar. Is it possible to isolate / reproduce your usage in an Example project which you can attach. It might help the Mapbox team to investigate it. I haven't seen the behavior inside a plain UIViewController but I'm also interested to debug it further in that particular situation.

**Comment by iamZoltanVaradi** [comment]
@mattijsf Forgot to mention I only experience this on iOS Simulator. I just went out and did a test on a real device, there the Map works great.

**Comment by jmkiley** [comment]
Hi @iamZoltanVaradi and @mattijsf -

Can you confirm which version of Xcode you are seeing this with? Since you are seeing the issue primarily on the iOS Simulator, it may be related to #9959.

**Comment by mattijsf** [comment]
Hi @jmkiley 

I repeated the same test with the same project on Xcode 9.1 with these (real) devices:
- iPhone 5 / iOS 10.3.3
- Iphone 5s / iOS 11.0.2

In the graphs below I first opened the app, scrolled continuously up/down while the map remained visible on screen. Then I scrolled down until the map disappears from the screen, again scrolled continuously up/down keeping the map off screen while doing it. On iOS 11 you'll see no difference, but on iOS 10 you'll see a clear spike when I started scrolling up / down off screen.

Unfortunately The results are very similar to my previous simulator tests:

**iOS 10 (Bad performance):**
![xcode91_ios10_3](https://user-images.githubusercontent.com/706368/32655400-a7b9510e-c60d-11e7-8d39-a24da2caed51.png)

**iOS 11: (Good performance):**
![xcode91_ios11](https://user-images.githubusercontent.com/706368/32655402-aa491ca6-c60d-11e7-9dde-2b3e7b1304b6.png)

**Comment by mattijsf** [comment]
I also did the same test with Xcode 9.1 iPhone 5s simulators running iOS 10.2 / iOS 11.1 but I got the same behavior / graphs from my opening comment.

It's nothing fancy but I uploaded my sample app here: https://github.com/mattijsf/MapboxTestApp

**Comment by mattijsf** [comment]
Is it possible to get some insight whether or not this is something that is being looked into? Might it be something that cannot be fixed at all? It's quite an impact on our iOS 10 userbase and I would love to know if this is heading somewhere.

**Comment by friedbunny** [comment]
As a workaround: If you don’t need interactive maps, I’d suggest checking out [`MGLSnapshotter`](https://www.mapbox.com/ios-sdk/api/3.7.1/Classes/MGLMapSnapshotter.html), available from v3.7.0.

**Comment by mattijsf** [comment]
Thanks @friedbunny. I'm not a fan because we use interactive maps for animating map content but it might be a direction we have to migrate towards.
